ansible/roles/docker/tasks/main.yml

```diff
@@ -41,26 +42,63 @@
   shell: systemctl status docker-tcp.socket | grep "Active:"
   register: docker_tcp_socket_state
 
+- include: create_docker_device.yml
+  when: docker_device != ""
+
 - name: copy systemd units for docker tcp socket settings
   template: src=docker-tcp.j2 dest=/etc/systemd/system/docker-tcp.socket
   register: docker_tcp_socket
 
   # tcp socket service requires docker service to be started after it
-- name: start docker tcp socket service
-  shell: sudo systemctl stop docker && sudo systemctl start docker-tcp.socket && sudo systemctl start docker
-  when: (docker_tcp_socket | changed) or not docker_tcp_socket_state.stdout | match("Active.* active")
+- name: reload systemd configuration
+  shell: sudo systemctl daemon-reload
+  when: "(docker_tcp_socket | changed) or (docker_tcp_socket_state.stdout != 'Active: active')"
+
+- name: stop docker
+  service:
+    name: docker
+    state: stopped
+  when: "(docker_tcp_socket | changed) or (docker_tcp_socket_state.stdout != 'Active: active')"
+
+- name: start docker-tcp service
+  service:
+    name: docker-tcp.socket
+    state: started
+  when: "(docker_tcp_socket | changed) or (docker_tcp_socket_state.stdout != 'Active: active')"
+
+- name: check docker service state
+  shell: sudo systemctl status docker | grep 'Active.*active' -o
+  ignore_errors: true
+  register: docker_service_state
+  tags:
+    - prebake-for-dev
 
 - name: remove the docker key file, if any. It shall be regenerated by docker on restart
   file: name=/etc/docker/key.json  state=absent
 
 # XXX: service module doesn't do daemon-reload yet, so need to use shell module here
 # https://github.com/ansible/ansible-modules-core/issues/191
-- name: restart docker
+- name: reload docker systemd configuration
   #service: name=docker state=restarted
-  shell: sudo systemctl daemon-reload && sudo systemctl restart docker
+  shell: sudo systemctl daemon-reload
+  when: "(docker_service_state.stderr | match('.*docker.service changed on disk.*')) or (docker_service_state.stdout != 'Active: active')"
   tags:
     - prebake-for-dev
 
+# XXX: this needs to happen twice after setting up the thin provisioner because
+#      of some docker bug I've not investigated.
+- name: restart docker (first time)
+  service:
+    name: docker
+    state: restarted
+  when: thin_provisioned
```

Comment:
@erikh this restarts docker on `make provision`, that will stop any running container. I think that is due to the fact that `thin_provisioned` is not set to `false` (as when the `lvcreate contiv-dockerthin` task is skipped it still ends up register this variable).

here is what I see, note that the `restart docker (first time)` task in both `make start` and `make provision` is triggered causing docker to be restarted kind of unconditionally:

## make start ##
```
TASK [docker : copy systemd units for docker tcp socket settings] **************
changed: [mon0]
changed: [mon1]
changed: [mon2]

TASK [docker : reload systemd configuration] ***********************************
changed: [mon0]
changed: [mon1]
changed: [mon2]

TASK [docker : stop docker] ****************************************************
ok: [mon0]
ok: [mon1]
ok: [mon2]

TASK [docker : start docker-tcp service] ***************************************
changed: [mon0]
changed: [mon1]
changed: [mon2]

TASK [docker : check docker service state] *************************************
changed: [mon2]
changed: [mon1]
changed: [mon0]

TASK [docker : remove the docker key file, if any. It shall be regenerated by docker on restart] ***
changed: [mon1]
changed: [mon0]
changed: [mon2]

TASK [docker : reload docker systemd configuration] ****************************
changed: [mon0]
changed: [mon1]
changed: [mon2]

TASK [docker : restart docker (first time)] ************************************
fatal: [mon1]: FAILED! => {"changed": false, "failed": true, "msg": "Warning: Stopping docker.service, but it can still be activated by:\n  docker-tcp.socket\nJob for docker.service failed because the control process exited with error code. See \"systemctl status docker.service\" and \"journalctl -xe\" for details.\n"}
...ignoring
fatal: [mon2]: FAILED! => {"changed": false, "failed": true, "msg": "Warning: Stopping docker.service, but it can still be activated by:\n  docker-tcp.socket\nJob for docker.service failed because the control process exited with error code. See \"systemctl status docker.service\" and \"journalctl -xe\" for details.\n"}
...ignoring
fatal: [mon0]: FAILED! => {"changed": false, "failed": true, "msg": "Warning: Stopping docker.service, but it can still be activated by:\n  docker-tcp.socket\nJob for docker.service failed because the control process exited with error code. See \"systemctl status docker.service\" and \"journalctl -xe\" for details.\n"}
...ignoring

TASK [docker : ensure docker is started] ***************************************
changed: [mon1]
changed: [mon0]
changed: [mon2]
```

## make provision ##
```
TASK [docker : lvcreate contiv-dockerthin] *************************************
skipping: [mon0]
skipping: [mon1]
skipping: [mon2]

TASK [docker : copy systemd units for docker tcp socket settings] **************
ok: [mon2]
ok: [mon1]
ok: [mon0]

TASK [docker : reload systemd configuration] ***********************************
changed: [mon1]
 [WARNING]: Consider using 'become', 'become_method', and 'become_user' rather
than running sudo

changed: [mon2]
changed: [mon0]

TASK [docker : stop docker] ****************************************************
ok: [mon1]
ok: [mon0]
ok: [mon2]

TASK [docker : start docker-tcp service] ***************************************
ok: [mon0]
ok: [mon1]
ok: [mon2]

TASK [docker : check docker service state] *************************************
fatal: [mon1]: FAILED! => {"changed": true, "cmd": "sudo systemctl status docker | grep 'Active.*active' -o", "delta": "0:00:00.020530", "end": "2016-03-01 12:07:15.903307", "failed": true, "rc": 1, "start": "2016-03-01 12:07:15.882777", "stderr": "", "stdout": "", "stdout_lines": [], "warnings": ["Consider using 'become', 'become_method', and 'become_user' rather than running sudo"]}
...ignoring
fatal: [mon0]: FAILED! => {"changed": true, "cmd": "sudo systemctl status docker | grep 'Active.*active' -o", "delta": "0:00:00.021246", "end": "2016-03-01 12:07:15.921782", "failed": true, "rc": 1, "start": "2016-03-01 12:07:15.900536", "stderr": "", "stdout": "", "stdout_lines": [], "warnings": ["Consider using 'become', 'become_method', and 'become_user' rather than running sudo"]}
...ignoring
fatal: [mon2]: FAILED! => {"changed": true, "cmd": "sudo systemctl status docker | grep 'Active.*active' -o", "delta": "0:00:00.032452", "end": "2016-03-01 12:07:15.957005", "failed": true, "rc": 1, "start": "2016-03-01 12:07:15.924553", "stderr": "", "stdout": "", "stdout_lines": [], "warnings": ["Consider using 'become', 'become_method', and 'become_user' rather than running sudo"]}
...ignoring

TASK [docker : remove the docker key file, if any. It shall be regenerated by docker on restart] ***
changed: [mon0]
changed: [mon1]
changed: [mon2]

TASK [docker : reload docker systemd configuration] ****************************
changed: [mon1]
changed: [mon0]
changed: [mon2]

TASK [docker : restart docker (first time)] ************************************
changed: [mon1]
changed: [mon0]
changed: [mon2]

TASK [docker : ensure docker is started] ***************************************
ok: [mon0]
ok: [mon2]
ok: [mon1]

```